File: servers/python/satellite_server/replica/manager.py
"""Replica manager — scheduled and on-demand sync from a remote primary satellite."""

from __future__ import annotations

import asyncio
import logging
import time
from collections.abc import Callable
from typing import Any

import httpx

from satellite_server.config.schema import CollectionConfig, SyncTrigger, WriteMode
from satellite_server.interfaces import IObjectStore
from satellite_server.protocol.push import push
from satellite_server.protocol.types import PushSuccess

logger = logging.getLogger(__name__)


def _deep_merge(local: dict[str, Any], remote: dict[str, Any]) -> dict[str, Any]:
    """Remote-wins deep merge.

    Recursively merges *remote* into *local*: for nested dicts both sides share,
    keys are merged recursively; for all other values, the remote value wins.
    This mirrors the ``defaultMerge`` behaviour in the TypeScript client SDK.
    """
    merged = {**local}
    for key, remote_val in remote.items():
        local_val = merged.get(key)
        if isinstance(remote_val, dict) and isinstance(local_val, dict):
            merged[key] = _deep_merge(local_val, remote_val)
        else:
            merged[key] = remote_val
    return merged


class ReplicaManager:
    """Manages replication from remote (primary) satellite servers.

    For each collection that has a ``remote`` field in its config, the
    ``ReplicaManager`` handles syncing data from the primary to local storage.
    Behaviour (write mode, sync triggers, interval) is fully driven by the
    collection config — no additional runtime configuration is needed.

    A single server can act as both primary and replica simultaneously.

    Typical usage::

        replica = ReplicaManager(store, config.collections,
                                 self_base_url="https://replica.example.com/v1")
        await replica.start()   # subscribe to primaries + begin scheduled tasks
        ...
        await replica.stop()    # clean shutdown

    The ``on_notification()`` and ``on_pull()`` methods are called by the
    replica router and pull route respectively.
    """

    def __init__(
        self,
        store: IObjectStore,
        collections: list[CollectionConfig],
        *,
        self_base_url: str | None = None,
        client: httpx.AsyncClient | None = None,
        on_error: Callable[[str, Exception], None] | None = None,
    ) -> None:
        self._store = store
        self._remote_cols = [c for c in collections if c.remote is not None]
        self._self_base_url = self_base_url
        self._owned_client = client is None
        self._client = client or httpx.AsyncClient(timeout=30.0)
        self._on_error = on_error or (
            lambda name, exc: logger.error("[ReplicaManager] %s: %s", name, exc)
        )
        # In-memory last-known hash per collection (avoids redundant writes)
        self._last_hash: dict[str, str] = {}
        # Monotonic timestamp (seconds) of the last completed sync per collection
        self._last_sync_at: dict[str, float] = {}
        self._tasks: list[asyncio.Task[None]] = []

    # ── Lifecycle ─────────────────────────────────────────────────────────────

    async def start(self) -> None:
        """Start background sync tasks and subscribe to primaries (webhook trigger).

        Calling ``start()`` multiple times is safe — any already-running tasks
        are left unchanged.
        """
        for col in self._remote_cols:
            remote = col.remote  # type: ignore[assignment]  # filtered above

            # Subscribe to primary so it notifies us on write
            if SyncTrigger.WEBHOOK in remote.sync_triggers and self._self_base_url:
                asyncio.create_task(self._subscribe(col))

            # Launch scheduled sync loop
            if SyncTrigger.SCHEDULED in remote.sync_triggers:
                task = asyncio.create_task(self._run_loop(col))
                self._tasks.append(task)
            else:
                # Even without a scheduled trigger, do one immediate sync on start
                asyncio.create_task(self._sync_safe(col))

    async def stop(self) -> None:
        """Cancel all background tasks and close the HTTP client (if owned)."""
        for task in self._tasks:
            task.cancel()
        if self._tasks:
            await asyncio.gather(*self._tasks, return_exceptions=True)
        self._tasks.clear()
        if self._owned_client:
            await self._client.aclose()

    # ── Trigger entry points ──────────────────────────────────────────────────

    async def on_notification(self, collection_name: str) -> None:
        """Called by ``POST /replica/notify`` when the primary signals a write."""
        col = self._find(collection_name)
        if col is None:
            logger.warning("[ReplicaManager] Notification for unknown collection %r", collection_name)
            return
        await self._sync_safe(col)

    async def on_pull(self, collection_name: str) -> None:
        """Called by the pull route when ``on_pull`` is listed in ``sync_triggers``.

        Awaited before the local store is read, ensuring the response is fresh.
        If ``on_pull_min_interval_ms`` is configured and the last sync occurred within
        that window, the primary is not contacted and cached local data is served instead.
        """
        col = self._find(collection_name)
        if col is None:
            return

        min_interval_ms = col.remote.on_pull_min_interval_ms if col.remote else None  # type: ignore[union-attr]
        if min_interval_ms is not None:
            last = self._last_sync_at.get(collection_name)
            if last is not None and (time.monotonic() - last) * 1000 < min_interval_ms:
                return  # within cooldown — serve cached local data

        await self._sync_safe(col)

    async def sync_now(self, name: str) -> None:
        """Trigger an immediate sync for a single collection by name."""
        col = self._find(name)
        if col is None:
            raise ValueError(f"[ReplicaManager] Unknown remote collection: {name!r}")
        await self._do_sync(col)

    async def sync_all(self) -> None:
        """Trigger an immediate sync for all remote collections in parallel."""
        await asyncio.gather(*(self._sync_safe(col) for col in self._remote_cols))

    # ── Internal ─────────────────────────────────────────────────────────────

    def _find(self, name: str) -> CollectionConfig | None:
        return next((c for c in self._remote_cols if c.name == name), None)

    async def _run_loop(self, col: CollectionConfig) -> None:
        interval = col.remote.interval_ms / 1000  # type: ignore[union-attr]
        while True:
            await self._sync_safe(col)
            await asyncio.sleep(interval)

    async def _sync_safe(self, col: CollectionConfig) -> None:
        try:
            await self._do_sync(col)
        except Exception as exc:  # noqa: BLE001
            self._on_error(col.name, exc)

    async def _do_sync(self, col: CollectionConfig) -> None:
        remote = col.remote  # type: ignore[assignment]
        document_key = col.storage_path  # static path — validated at config load time

        # Full pull from primary (no checkpoint — always get the authoritative full document)
        primary_url = f"{remote.url.rstrip('/')}{remote.pull_path}"
        resp = await self._client.get(
            primary_url,
            headers={"Accept": "application/json", **remote.headers},
        )
        resp.raise_for_status()
        pulled: dict[str, Any] = resp.json()

        primary_hash: str = pulled.get("hash", "")
        primary_data: dict[str, Any] = pulled.get("data", {})

        # Nothing on primary yet
        if not primary_hash:
            return

        # Skip write if primary hasn't changed since our last sync
        if self._last_hash.get(col.name) == primary_hash:
            return

        # Read the current local document to get its hash
        raw_local = await self._store.get_string(document_key)
        current_local_hash: str = ""
        current_local_data: dict[str, Any] = {}
        if raw_local:
            import json
            local_doc = json.loads(raw_local)
            current_local_hash = local_doc.get("hash", "")
            current_local_data = local_doc.get("data", {})

        # Local store already matches primary — just update in-memory state
        if current_local_hash == primary_hash:
            self._last_hash[col.name] = primary_hash
            return

        # Determine data to write based on write mode
        if remote.write_mode == WriteMode.BIDIRECTIONAL and current_local_data:
            data_to_write = _deep_merge(current_local_data, primary_data)
        else:
            # PULL_ONLY and WRITE_THROUGH: mirror primary exactly
            data_to_write = primary_data

        # Write to local store using the push protocol
        # base_hash=None → first write (doc doesn't exist); otherwise use current local hash
        base_hash = current_local_hash if current_local_hash else None
        result = await push(self._store, document_key, data_to_write, base_hash)

        if not isinstance(result, PushSuccess):
            # Concurrent write between our read and write — self-corrects on next sync
            raise RuntimeError(
                f"[ReplicaManager] Concurrent write on {col.name!r} — will retry"
            )

        self._last_hash[col.name] = result.hash
        self._last_sync_at[col.name] = time.monotonic()
        logger.debug("[ReplicaManager] Synced %r (hash=%s)", col.name, result.hash)

    async def _subscribe(self, col: CollectionConfig) -> None:
        """Register this replica's webhook URL with the primary."""
        remote = col.remote  # type: ignore[assignment]
        webhook_url = self._self_base_url
        subscribe_url = f"{remote.url.rstrip('/')}/replica/subscribe"
        try:
            resp = await self._client.post(
                subscribe_url,
                json={"webhook_url": webhook_url, "collections": [col.name]},
                headers={"Content-Type": "application/json", **remote.headers},
            )
            if not resp.is_success:
                logger.warning(
                    "[ReplicaManager] Subscription to %s returned HTTP %s",
                    subscribe_url,
                    resp.status_code,
                )
        except Exception as exc:  # noqa: BLE001
            logger.warning(
                "[ReplicaManager] Could not subscribe to %s: %s (scheduled pulls will still work)",
                subscribe_url,
                exc,
            )

File: servers/ts/src/replica/manager.ts
import { push } from "../protocol/index.js"
import type { IObjectStore } from "../interfaces.js"
import type { WildcardRemoteConfig } from "../config/schema.js"

/**
 * Manages replica-side synchronisation from a primary satellite server.
 *
 * Currently implements **wildcard on-demand pull**: any collection path not
 * matched by an explicit ``CollectionConfig`` can be fetched from the primary
 * on demand via {@link onPullWildcard}, cached locally, and served by the
 * wildcard catch-all pull route.
 *
 * Example:
 * ```ts
 * const manager = new ReplicaManager(store, {
 *   wildcardRemote: {
 *     url: "https://primary.example.com/v1",
 *     pullPathTemplate: "/pull/{name}",
 *     readRoles: ["public"],
 *   },
 * })
 * ```
 */
export class ReplicaManager {
  private readonly negativeCache = new Map<string, number>()     // path → Date.now()
  private readonly wildcardLastSyncAt = new Map<string, number>() // path → Date.now()
  private readonly wildcardLastHash = new Map<string, string>()   // path → hash

  constructor(
    private readonly store: IObjectStore,
    private readonly options: { wildcardRemote?: WildcardRemoteConfig } = {},
  ) {}

  /**
   * Called by the wildcard catch-all pull route for paths not matched by any
   * explicit collection.
   *
   * Fetches from the primary on demand, stores the result locally, and returns
   * `true` if data is available (freshly fetched or stale cached).
   *
   * Returns `false` when the collection does not exist or is not authorised on
   * the primary **and** no stale copy is held locally.
   *
   * Negative results (404 / 403 / 401) are cached for
   * `wildcardRemote.negativeCacheMs` to avoid hammering the primary. Stale
   * local data is still served during this window. On primary unavailability
   * (network errors, 5xx) stale local data is served when available.
   */
  async onPullWildcard(path: string): Promise<boolean> {
    const wc = this.options.wildcardRemote
    if (!wc) return false

    const now = Date.now()

    // ── Negative cache ────────────────────────────────────────────────────
    const negTs = this.negativeCache.get(path)
    if (negTs !== undefined) {
      if (now - negTs < wc.negativeCacheMs) {
        return (await this.store.getString(path)) !== null
      }
      // Cache expired — remove and retry
      this.negativeCache.delete(path)
    }

    // ── on-pull cooldown ──────────────────────────────────────────────────
    if (wc.onPullMinIntervalMs !== undefined) {
      const last = this.wildcardLastSyncAt.get(path)
      if (last !== undefined && now - last < wc.onPullMinIntervalMs) {
        const local = await this.store.getString(path)
        if (local !== null) return true
        // No local data despite being within cooldown — fall through to fetch
      }
    }

    // ── Fetch from primary ────────────────────────────────────────────────
    const pullPath = wc.pullPathTemplate.replace("{name}", path)
    const primaryUrl = `${wc.url.replace(/\/$/, "")}${pullPath}`

    let resp: Response
    try {
      resp = await fetch(primaryUrl, {
        headers: { Accept: "application/json", ...wc.headers },
      })
    } catch (err) {
      console.warn(`[ReplicaManager] Wildcard fetch failed for "${path}":`, err)
      return (await this.store.getString(path)) !== null
    }

    if (resp.status === 401 || resp.status === 403 || resp.status === 404) {
      this.negativeCache.set(path, now)
      console.debug(
        `[ReplicaManager] Wildcard "${path}": primary returned HTTP ${resp.status} — negative-cached`,
      )
      return (await this.store.getString(path)) !== null
    }

    if (!resp.ok) {
      console.warn(
        `[ReplicaManager] Wildcard fetch for "${path}" returned HTTP ${resp.status} — serving stale`,
      )
      return (await this.store.getString(path)) !== null
    }

    const pulled = (await resp.json()) as { data?: Record<string, unknown>; hash?: string }
    const primaryHash = pulled.hash ?? ""
    const primaryData = pulled.data ?? {}

    if (!primaryHash) {
      // Empty collection on primary
      return false
    }

    // Skip write if we already have this version
    if (this.wildcardLastHash.get(path) !== primaryHash) {
      const rawLocal = await this.store.getString(path)
      let currentHash: string | null = null
      if (rawLocal !== null) {
        currentHash = (JSON.parse(rawLocal) as { hash?: string }).hash ?? null
      }

      if (currentHash !== primaryHash) {
        const result = await push(this.store, path, primaryData, currentHash)
        if (result.ok) {
          this.wildcardLastHash.set(path, result.hash)
        }
      } else {
        this.wildcardLastHash.set(path, primaryHash)
      }
    }

    this.wildcardLastSyncAt.set(path, now)
    console.debug(`[ReplicaManager] Wildcard synced "${path}" (hash=${primaryHash})`)
    return true
  }
}
